auto-tagging for all resources

scotwk opened the pull request:
@kapilt - I just want to confirm that I am doing the right thing.  I need to look up all resources that support tags and add in this?

```
actions.register('auto-tag-user', AutoTagUser)
```

So the work is in determining which resources support tags?

Issue #1296

kapilt commented:
i'd like to get the registries doing event notification and subscription on type registration, so generics like this or cloudwatch can be introspect and attach to relevant types. right now cloudwatch is using the resource metaclass to watch at class construction time.

scotwk commented:
@kapilt - do you know what I introspect for tags or is that what I have to figure out?  I was looking at the boto3 library to see if I could determine which resources tagging supports, but I did not see an obvious sentinel to indicate tagging (I was looking for a consistent `create_tags` method which was not present for some resources)

alfredgamulo commented:
@scotwk I'm working on a PR now that I think fulfills what you're trying to do. I'm working on specifically adding this for cloudfront and vpc though as per my need.

kapilt commented:
re the registry subscriber, also need to introduce post/final style events, else too many ordering concerns.

kapilt commented:
@scotwk introspection should be on presence of tag action to add autotag.

scotwk commented:
@kapilt - is this what you were thinking?  I added in the ability for events (aka hooks) into the registry.  The only one being used right now is the 'final' hook, but I wanted to make it extensible so I also included an event during `register` and `unregister`.  These can be removed easily if you think its YAGNI.

The one hook I added will run after all actions are registered and it adds in the `auto-tag-user` action to anything with a `tag` action

This is going to have conflicts with #1315 but they should be easily resolved (I'm thinking it will be mainly around the fact that we both removed the AutoTagUser stuff from each of the resources using it).

kapilt commented on c7n/registry.py:
```diff
@@ -45,28 +45,47 @@ class PluginRegistry(object):
         self._factories = {}
+        self._events = {x: [] for x in self.event_times}
+
+    def add_event(self, time, func):
```
subscribe is a bit more clear, also time event_name/event_type.. time is ambiguous.

kapilt commented on c7n/registry.py:
```diff
@@ -45,28 +45,47 @@ class PluginRegistry(object):
 
     """
 
+    event_times = ('register', 'unregister', 'final')
```
not sure we need an unregister

kapilt commented on c7n/registry.py:
```diff
@@ -45,28 +45,47 @@ class PluginRegistry(object):
+    def run_final_events(self):
+        self.run_events('final', None)
+
+    def run_events(self, category, key):
```
notify re name .. run_events is ambiguous.

kapilt commented on c7n/registry.py:
```diff
@@ -45,28 +45,47 @@ class PluginRegistry(object):
+
+    def run_events(self, category, key):
+        for event in self._events[category]:
+            event(self, key, self._factories.keys())
```
s/event/subscriber

also we should be distinguishing on collection vs registration events on params. registratin events should get (event_name, registry_key, and value) .. collection/registry events should get (event_name, registry).

kapilt commented on c7n/resources/__init__.py:
```diff
@@ -76,3 +76,7 @@ def load_resources():
     from c7n.manager import resources
     resources.load_plugins()
+
+    for _, klass in resources.items():
```
rather than run events per Klass, action/filter registry. run once for the resource type registry.

kapilt commented on c7n/registry.py:
```diff
@@ -45,46 +45,45 @@ class PluginRegistry(object):
-        self.run_events('final', None)
+    def notify(self, category, key):
+        for subscriber in self._subscribers[category]:
+            subscriber(self, key, self._factories.keys())
```
drop the third parameter here, if a subscriber wants the keys it can use the registry methods. if the registry needs keys() method then lets add that.

kapilt commented on c7n/registry.py:
```diff
@@ -45,46 +45,45 @@ class PluginRegistry(object):
     """
 
-    event_times = ('register', 'unregister', 'final')
+    events = ('register', 'final')
```
let's make these values class constants so subscribers aren't using string values.

kapilt commented on c7n/registry.py:
```diff
@@ -45,46 +45,45 @@ class PluginRegistry(object):
-    def run_events(self, category, key):
-        for event in self._events[category]:
-            event(self, key, self._factories.keys())
+    def notify_loading_done(self):
```
let's ditch this helper method, and direct call notify.

scotwk commented:
@kapilt - thanks for the reviews.  I think it's good to go.

kapilt commented on c7n/registry.py:
```diff
@@ -45,19 +45,32 @@ class PluginRegistry(object):
     def register(self, name, klass=None):
         # invoked as function
         if klass:
+            self.notify(self.EVENT_REGISTER, name)
```
i'd suggest modifying the param here to be k/class and do it a few lines laster.

kapilt commented on c7n/manager.py:
```diff
@@ -16,12 +16,21 @@
 from c7n.utils import dumps
 
 
+def add_auto_tag_user(registry, _):
```
i'd suggest moving this to near the auto tags definition,  else its a bit of action at a distance.

JohnTheodore commented:
Here is a list of resources that support tagging: https://docs.aws.amazon.com/awsconsolehelpdocs/latest/gsg/supported-resources.html#supported-resources-tagging-console

JohnTheodore commented:
@scotwk I pulled down this PR, the auto-tag-user didn't work for rds-snapshot or app-elb

scotwk commented:
@JohnTheodore - what issue are you seeing?  

```
$ custodian schema rds-snapshot.actions
load resources
rds-snapshot:
  actions: [auto-tag-user, delete, invoke-lambda, mark-for-op, notify, put-metric,
    region-copy, remove-tag, restore, tag, unmark]

$ custodian schema app-elb.actions
load resources
app-elb:
  actions: [auto-tag-user, delete, invoke-lambda, mark-for-op, modify-listener, notify,
    put-metric, remove-tag, set-s3-logging, set-shield, set-waf, tag]
```

scotwk commented:
@kapilt - This is ready to go.  I have resolved the merge conflicts with master and also handled the two comments from your last review.  This is a few months old, so perhaps a new review is in order since the surrounding code has changed a bit.

JohnTheodore commented:
I pulled the latest code and rds-snapshot works great with auto-tag-user now.

I'll test several other services and message this thread with any issues. thanks for the great PR, this makes my life easier!

kapilt commented on c7n/manager.py:
```diff
@@ -16,12 +16,14 @@
 
 
 resources = PluginRegistry('resources')
+resources.subscribe(resources.EVENT_FINAL, add_auto_tag_user)
```
i'd rather have the subscription next to the subscriber ie move this to actions.py, its clearer on usage of the subscriber, and avoids potential import cycles, manager is effectively the root of the class tree.

scotwk commented:
@kapilt - done

FireballDWF commented:
Request reopening as the following resources are taggable but custodian is not supporting tagging:
1. waf
2. waf-regional
3. Elasticsearch (aka es)

scotwk commented:
@FireballDWF - this PR was to add auto-tag-user to anything that supports the tag action.  It was not to add the tag action everywhere.  I think it would be more appropriate to open a new issue for any resources missing the tag action
